src/hunk.rs

```diff
@@ -0,0 +1,104 @@
+use std::ops::Range;
+
+use anyhow::{anyhow, bail, Result};
+use line_span::{LineSpan, LineSpans};
+
+pub struct Hunk {
+    /// Raw hunk string.
+    ///
+    /// This may not be modified after the hunk was created, as the other fields index into it.
+    raw_hunk: String,
+    /// Header marker and content.
+    header: (Range<usize>, Range<usize>),
+    /// Ranges of `raw_hunk` for each line with the corresponding diff type.
+    ///
+    /// Doesn't include the starting character of the diff line (`+/-/ `) or the line break.
+    lines: Vec<HunkLine>,
+}
+
+impl Hunk {
+    pub fn from_string(raw_hunk: String) -> Result<Self> {
+        let mut lines_iter = raw_hunk.line_spans();
+        let header = match lines_iter.next().map(Self::parse_header) {
+            Some(Ok(header)) => header,
+            Some(Err(e)) => return Err(e),
+            _ => bail!("Empty hunk."),
+        };
+        let lines = lines_iter
+            .map(HunkLine::from_line_span)
+            .collect::<Result<Vec<_>, _>>()?;
+        Ok(Self {
+            raw_hunk,
+            header,
+            lines,
+        })
+    }
+
+    pub fn raw(&self) -> &str {
+        &self.raw_hunk
+    }
+
+    pub fn header(&self) -> (&str, &str) {
+        (
+            &self.raw_hunk[self.header.0.clone()],
+            &self.raw_hunk[self.header.1.clone()],
+        )
+    }
+
+    pub fn lines(&self) -> impl Iterator<Item = (DiffLineType, &str)> {
+        self.lines
+            .iter()
+            .map(|line| (line.diff_type, &self.raw_hunk[line.range.clone()]))
+    }
+
+    fn parse_header(line: LineSpan) -> Result<(Range<usize>, Range<usize>)> {
+        // TODO(cptp)
+        let range = line.range();
+        Ok((range.start..range.start, range))
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum DiffLineType {
+    Unchanged,
```

Comment:
Bikeshed (not important): how do we feel about changing this variant to `Context` to bring it inline with git's documentation?

Reply:
Sounds good, I'm not familiar with the exact naming in git so I just chose something. I'll also use "deleted" instead of "removed" since I think I saw these used in other parts of gex.

Reply:
I didn't pay it much thought before but come to think of it, git actually uses "old" and "new" where in gex we use "deletion" and "addition", so it's already inconsistent. To be honest old and new don't sound as obvious to me personally, could just be my bias.

Anyway, this is far from an important point about the code so I will stop fixating on it :sweat_smile: